File: benchmark/autolifter/single-pass/average.f
Config ExtraGrammar = "AutoLifter";

Inductive List = cons {Int, List} | nil Unit;

div = \a: Int. \b: Int.
  if == b 0 then 0 else / a b;

length = fix (
  \f: List -> Int. \xs: List.
  match xs with
    nil _ -> 0
  | cons {h, t} -> + 1 (f t)
  end
);

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

average = \xs: List.
  let len = length xs in
  let sum = fix (
    \f: List -> Int. \ys: List.
    match ys with
      nil _ -> 0
    | cons {h, t} -> + h (f t)
    end
  ) xs in div sum len;

main = single_pass average;

File: ml-benchmark/autolifter/single-pass/average.ml
type list_ = Cons of int * list_ | Nil of unit

let div a b = if b = 0 then 0 else a / b

let rec length xs =
  match xs with
  | Nil _ -> 0
  | Cons (h, t) -> 1 + length t

let single_pass v =
  let rec run xs =
    match xs with
    | Nil _ -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let average xs =
  let len = length xs in
  let rec f ys =
    match ys with
    | Nil _ -> 0
    | Cons (h, t) -> h + f t
  in
  let sum = f xs in
  div sum len

let program = single_pass average